src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt

```diff
@@ -98,104 +235,213 @@ open class GossipRouter : AbstractRouter() {
+                    if (scoreMedian < scoreParams.opportunisticGraftThreshold) {
+                        (getTopicPeers(topic) - peers)
+                            .filter { score.score(it) > scoreMedian && !isDirect(it) && !isBackOff(it, topic) }
+                            .take(params.opportunisticGraftPeers)
+                            .forEach { graft(it, topic) }
                     }
                 }
-                submitGossip(topic, peers)
+
+                emitGossip(topic, peers)
             }
             fanout.entries.forEach { (topic, peers) ->
-                peers.removeIf { it in getTopicPeers(topic) }
-                val needMore = D - peers.size
+                peers.removeIf {
+                    it !in getTopicPeers(topic) || score.score(it) < scoreParams.publishThreshold
+                }
+                val needMore = coreParams.D - peers.size
                 if (needMore > 0) {
-                    peers += (getTopicPeers(topic) - peers).shuffled(random).take(needMore)
+                    peers += (getTopicPeers(topic) - peers)
+                        .filter { score.score(it) >= scoreParams.publishThreshold && !isDirect(it) }
+                        .shuffled(random)
+                        .take(needMore)
                 }
-                submitGossip(topic, peers)
+                emitGossip(topic, peers)
             }
             lastPublished.entries.removeIf { (topic, lastPub) ->
-                (time - lastPub > fanoutTTL)
+                (curTime() - lastPub > coreParams.fanoutTTL.toMillis())
                     .whenTrue { fanout.remove(topic) }
             }
 
-            (mesh.keys.toSet() + fanout.keys).forEach { topic ->
-                val gossipPeers = (getTopicPeers(topic) - mesh[topic]!! - (fanout[topic] ?: emptyList()))
-                    .shuffled(random).take(DGossip)
-                submitGossip(topic, gossipPeers)
-            }
-
             mCache.shift()
+
             flushAllPending()
         } catch (t: Exception) {
             logger.warn("Exception in gossipsub heartbeat", t)
         }
     }
 
-    private fun prune(peer: PeerHandler, topic: String) = addPendingRpcPart(
-        peer,
-        Rpc.RPC.newBuilder().setControl(
-            Rpc.ControlMessage.newBuilder().addPrune(
-                Rpc.ControlPrune.newBuilder().setTopicID(topic)
-            )
-        ).build()
-    )
+    private fun emitGossip(topic: Topic, excludePeers: Collection<PeerHandler>) {
+        val ids = mCache.getMessageIds(topic)
+        if (ids.isEmpty()) return
+
+        val shuffledMessageIds = ids.shuffled(random).take(params.maxIHaveLength)
+        val peers = (getTopicPeers(topic) - excludePeers)
+            .filter { score.score(it) >= score.params.gossipThreshold && !isDirect(it) }
+
+        peers.shuffled(random)
+            .take(max((params.gossipFactor * peers.size).toInt(), coreParams.DLazy))
```

Comment:
I think this should be:
```suggestion
            .take(min((params.gossipFactor * peers.size).toInt(), coreParams.DLazy))
```
Since the spec says:
> From these peers, it randomly selects gossip factor peers with a minimum of D_lazy, and emits gossip to the selected peers.

Reply:
And it might be a good idea to add a test for it

Reply:
From my understanding of the spec `D_lazy` is the lower bound so if the `params.gossipFactor * peers.size` value is lower than `D_lazy` then the `D_lazy` value should be used. This is `max`. 
May be I missed something?

Reply:
Yes, you're right. Sorry for the confusion.